Fix local solve

atgeirr opened the pull request:
This fixes behaviour for local solves. It combines two separate fixes:
 - ensure zero-flux connections are not included
 - take tracer < 1 into account for time-of-flight.

atgeirr committed:
Add clarifying comment.

atgeirr committed:
Only add connection if flux strictly different from zero.

While not a bug as such, adding zero fluxes has the following
unfortunate consequences:
 - creates unnecessary connected components (triggering unnecessary multi-cell solves)
 - causes the domains of local solutions to include unnecessary cells and components

The second is the most problematic, although filtering on zero tracer values could also
deal with it. The current fix is preferred though, as it also improves performance and
makes the algorithms perform more intuitively (i.e. not visiting unnecessary cells).

atgeirr committed:
Add slightly more complex test case.

Local TOF commented out as that has a bug that is currently unfixed.

atgeirr committed:
Bugfix: ensure local solves account for tracer < 1.

Pore volume and fluxes available to the tracer solved for
is reduced in the presence of outside tracers, leading to
differences in tof.

atgeirr committed:
Add tests for local tof that used to fail.

bska commented on opm/flowdiagnostics/TracerTofSolver.cpp:
```diff
@@ -305,13 +305,14 @@ namespace FlowDiagnostics
             const int upwind_cell = conn.neighbour;
             const double flux = conn.weight;
-            upwind_tof_contrib += tof_[upwind_cell] * flux;
+            upwind_tof_contrib += tof_[upwind_cell] * (tracer_[upwind_cell] != 0.0 ? tracer_[upwind_cell] : 1.0) * flux;
```
Do we ever compute negative tracer concentrations?  If not, then this condition should probably be
```C++
tracer_[upwind_cell] > 0.0
```

atgeirr replied:
OK, will change.

bska commented on opm/flowdiagnostics/TracerTofSolver.cpp:
```diff
@@ -305,13 +305,14 @@ namespace FlowDiagnostics
         // Compute time-of-flight and tracer.
-        tof_[cell] = (pv_[cell] + upwind_tof_contrib) / total_influx;
         tracer_[cell] = is_start_[cell] ? 1.0 : upwind_tracer_contrib / total_influx;
+        const double tracer = tracer_[cell] != 0.0 ? tracer_[cell] : 1.0;
```
Same here.

bska commented on tests/test_flowdiagnosticstool.cpp:
```diff
@@ -40,6 +40,7 @@
 #include <opm/utility/numeric/RandomVector.hpp>
 
 #include <algorithm>
+#include <iostream>
```
Preferably not.

atgeirr replied:
Will replace with boost test messages.

bska reviewed (commented):
I suppose it's okay, but I have a couple of comments regarding the details.  Maybe there's even a better way of accomplishing the goal.

atgeirr committed:
Revise according to review input and other improvements.

In particular:
 - Remove start sets argument for TracerTofSolver::solveGlobal(),
   using source term as start indicator instead.
 - Use tracer info more elegantly in solveSingleCell(), no arithmetic ifs.
 - Revise test that was wrong (corner case: global tof should not be affected
   by start sets) after the solveGlobal() change.
 - Use boost test messages instead of iostreams.

atgeirr commented:
The PR has been updated and improved.

bska commented on tests/test_flowdiagnosticstool.cpp:
```diff
@@ -439,4 +439,198 @@ BOOST_AUTO_TEST_CASE (OneDimCase)
+                                        {
+                                            return s.id().to_string() == pt.to_string();
+                                        });
+                BOOST_CHECK(pos != injstart.end());
```
Did you mean
```C++
BOOST_CHECK(pos != s1.end());
```
here?  Otherwise, I don't understand how the check could be meaningful for `pairs[1]`.

atgeirr replied:
You are correct, a bugfix has been pushed.

bska reviewed (commented):
I'm getting errors in a checked iterator build of this when running `make test`.  I think there is an error in the check that verifies the set of start points.

atgeirr committed:
Bugfix: use correct container.

bska commented:
> a bugfix has been pushed.

Thanks.  I'll merge into master.
